src/shell/tmpdir.c

```diff
@@ -0,0 +1,102 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "src/common/libutil/cleanup.h"
+
+#include "builtins.h"
+#include "internal.h"
+#include "info.h"
+
+#define JOBTMP_SIZE 1024
+
+static int mkdir_exist_ok (const char *path)
+{
+    if (mkdir (path, 0700) < 0) {
+        if (errno != EEXIST) {
+            shell_log_errno ("mkdir %s", path);
+            return -1;
+        }
+    }
+    return 0;
+}
+
+static int make_job_path (const char *parent,
+                          flux_shell_t *shell,
+                          char *buf,
+                          size_t size)
+{
+    size_t n;
+
+    n = snprintf (buf, size, "%s/jobtmp-%d-", parent, shell->info->shell_rank);
```

Comment:
Smart idea to put the shell rank in the directory name!

Reply:
Thanks - the idea was to let each shell manage one directory even when there are multple shells/brokers per node, and thus not clobber each other.  However in testing on opal I'm running into trouble when shells span nodes.  I think I have a flawed assumption that that FLUX_JOB_TMPDIR would be inherited from the enclosing instance _shell_ on a given node, not from the submitting node.

Reply:
Er, above I said I was testing on opal, but it's really quartz.

Reply:
Ah, glad you caught that.

Reply:
Hmm, thought I could solve this by getting FLUX_JOB_TMPDIR using `getenv()` instead of `flux_shell_getenv()` but I think that returns the value from the submitting broker's environment, not the executing one.

Beginning to wonder if it is a good idea to try to have the directory hierarchy track the job hierarchy.  The problem gets a bit simpler if we just create these directories in a flat namespace under `${TMPDIR:-/tmp}`.  But then cleanup is harder.

Reply:
It is nice if each shell has its own FLUX_JOB_TMPDIR when the shells share a node, but what is the fallout if we instead use the same FLUX_JOB_TMPDIR per-_job_ (as I suppose the name does imply)? We're still better off than with no per-job tmpdir...

Reply:
> but I think that returns the value from the submitting broker's environment, not the executing one.

Currently I think it would always be the rank 0 broker's environment. IIRC, the job-exec module copies the current broker's environment when creating the `flux_cmd_t` for the shell (or the IMP in the case of multi-user). Eventually the local broker MAY directly execute the job shell, but in the case of multiuser, I think extra environment sanitization is done before exec(), so I don't think any assumptions can be made about environment inheritance in the shell itself.

Worst case, if we really need that shell rank in the FLUX_JOB_TMPDIR, maybe the path can be set as a broker attribute (somehow)?

Reply:
> but what is the fallout if we instead use the same FLUX_JOB_TMPDIR per-job

Well cleanup could be a bit racy with multiple brokers per node, at least as I envisioned it:  each shell would do a mkdir and ignore EEXIST errors.  Then each shell would set up an atexit handler to `rm -r` the directory.  If one shell exits early, then another shell might find its directory gone.  I worry that since this is an open-ended feature, that this might result in some hard to diagnose failure modes down the road.

I suppose we could go with a single directory that we don't try to clean up in the shell, and then leave it to `flux-start` or a system instance epilogue to do the cleanup (cleaning up the job it launched and any sub-jobs contained within).

Reply:
>  I worry that since this is an open-ended feature, that this might result in some hard to diagnose failure modes down the road.

It could be misguided to bend over backwards to support the multiple job-shells per node, though I guess that does drive most of our tests (counterpoint: issues here would be discovered early. counter-counterpoint: I hate getting off in the weeds fixing testing-only issues)

What about a broker attribute then?  It might actually be useful to be able to query the current brokers idea of FLUX_JOB_TMPDIR (e.g. your flux-proxy case)

> multi-user, where a job might not have write permission in its parents jobtmp.

Yeah, I agree FLUX_JOB_TMPDIR should not be inherited over userid transitions.